### ISSUE TITLE
Bad alias info when trying to assign into a "mutable static array"?

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/julia#37780]
TLDR: When doing a matrix multiply into a mutable static matrix, the whole inner tuple gets reloaded in the inner iteration, leading to bad performance.

The code below contains a mini version of a 4x4 Float64 static matrix (one immutable and one mutable version). It also contains a matrix multiplication function.

```jl
abstract type StaticMatrix4x4F64 <: AbstractMatrix{Float64} end
Base.@propagate_inbounds Base.getindex(v::StaticMatrix4x4F64, i::Int) = v.x[i]
Base.IndexStyle(::Type{<:StaticMatrix4x4F64}) = IndexLinear()
Base.size(::StaticMatrix4x4F64) = (4,4)

struct SMatrix4x4F64 <: StaticMatrix4x4F64
    x::NTuple{16, Float64}
end

mutable struct MMatrix4x4F64 <: StaticMatrix4x4F64
    x::NTuple{16, Float64}
end

function Base.setindex!(v::MMatrix4x4F64, val, i::Int)
    T = eltype(v)
    @assert isbitstype(T)
    @boundscheck checkbounds(v,i)
    GC.@preserve v unsafe_store!(Base.unsafe_convert(Ptr{T}, pointer_from_objref(v)), convert(T, val), i)
    return v
end

function mymul!(c::MMatrix4x4F64, a::SMatrix4x4F64, b::SMatrix4x4F64) 
    I, J, K = 4, 4, 4
    @inbounds for k in 1:K, j in 1:J
        b_jk = b[j, k]
        for i in 1:I
            c[i, k] += a[i, j] * b_jk
        end
    end
    return c
end
```

We can benchmark this:

```jl
s = SMatrix4x4F64(ntuple(i -> Float64(i), 16))
m = MMatrix4x4F64(ntuple(i -> Float64(i), 16))
using BenchmarkTools
@btime mymul!($m, $s, $s)
#  345.529 ns (0 allocations: 0 bytes)
```

~350 ns for a 4x4 matrix multiply is incredibly slow so something must be going wrong. Looking at a snippet of the LLVM code (in the inner iteration):

```
julia> @code_llvm mymul!(m,s,s)
... # snip
L33:                                              ; preds = %L33, %L5
  %value_phi4 = phi i64 [ 1, %L5 ], [ %39, %L33 ]
; └
;  @ REPL[8]:6 within `mymul!'
; ┌ @ abstractarray.jl:1120 within `getindex'
; │┌ @ abstractarray.jl:1147 within `_getindex'
; ││┌ @ REPL[2]:1 within `getindex'
; │││┌ @ Base.jl:33 within `getproperty'
      %.unpack67 = load i64, i64* %5, align 8
      %17 = bitcast double* %.elt22 to <4 x i64>*
      %18 = load <4 x i64>, <4 x i64>* %17, align 8
      %19 = bitcast double* %.elt30 to <4 x i64>*
      %20 = load <4 x i64>, <4 x i64>* %19, align 8
      %21 = bitcast double* %.elt38 to <4 x i64>*
      %22 = load <4 x i64>, <4 x i64>* %21, align 8
      %23 = bitcast double* %.elt46 to <2 x i64>*
      %24 = load <2 x i64>, <2 x i64>* %23, align 8
      %.unpack5152 = load i64, i64* %6, align 8
      store i64 %.unpack67, i64* %7, align 8
      %25 = bitcast double* %.fca.1.gep to <4 x i64>*
      store <4 x i64> %18, <4 x i64>* %25, align 8
      %26 = bitcast double* %.fca.5.gep to <4 x i64>*
      store <4 x i64> %20, <4 x i64>* %26, align 8
      %27 = bitcast double* %.fca.9.gep to <4 x i64>*
      store <4 x i64> %22, <4 x i64>* %27, align 8
      %28 = bitcast double* %.fca.13.gep to <2 x i64>*
      store <2 x i64> %24, <2 x i64>* %28, align 8
      store i64 %.unpack5152, i64* %8, align 8
; │││└
; │││ @ REPL[2]:1 within `getindex' @ tuple.jl:29
     %29 = add nsw i64 %11, %value_phi4
     %30 = getelementptr inbounds [16 x double], [16 x double]* %3, i64 0, i64 %29
     %31 = add nsw i64 %15, %value_phi4
     %32 = getelementptr inbounds [1 x [16 x double]], [1 x [16 x double]]* %1, i64 0, i64 0, i64 %31
; └└└
; ┌ @ float.jl:405 within `*'
   %33 = load double, double* %32, align 8
   %34 = fmul double %33, %16
; └
; ┌ @ float.jl:401 within `+'
   %35 = load double, double* %30, align 8
   %36 = fadd double %35, %34
; └
; ┌ @ abstractarray.jl:1217 within `setindex!'
; │┌ @ abstractarray.jl:1240 within `_setindex!'
; ││┌ @ REPL[7]:5 within `setindex!'
; │││┌ @ pointer.jl:118 within `unsafe_store!'
      %37 = getelementptr inbounds double, double* %9, i64 %29
      store double %36, double* %37, align 1
; └└└└
; ┌ @ range.jl:627 within `iterate'
; │┌ @ promotion.jl:408 within `=='
    %38 = icmp eq i64 %value_phi4, 4
; │└
   %39 = add nuw nsw i64 %value_phi4, 1
; └
  br i1 %38, label %L137, label %L33
... # snap
```

Or the native code:

```
; │┌ @ abstractarray.jl:1120 within `getindex'
; ││┌ @ abstractarray.jl:1147 within `_getindex'
; │││┌ @ REPL[2]:1 within `getindex'
; ││││┌ @ Base.jl:33 within `getproperty'
L80:
        movq    (%rax), %rcx
        vmovups 8(%rax), %ymm1
        vmovups 40(%rax), %ymm2
        vmovups 72(%rax), %ymm3
        vmovups 104(%rax), %xmm4
        movq    120(%rax), %rdi
        movq    %rcx, -128(%rsp)
        vmovups %ymm1, -120(%rsp)
        vmovups %ymm2, -88(%rsp)
        vmovups %ymm3, -56(%rsp)
        vmovups %xmm4, -24(%rsp)
        movq    %rdi, -8(%rsp)
; │└└└└
; │┌ @ float.jl:405 within `*'
        vmulsd  (%r15,%rbx,8), %xmm0, %xmm1
; │└
; │┌ @ float.jl:401 within `+'
        vaddsd  (%r10,%rbx,8), %xmm1, %xmm1
; │└
; │┌ @ abstractarray.jl:1217 within `setindex!'
; ││┌ @ abstractarray.jl:1240 within `_setindex!'
; │││┌ @ REPL[7]:5 within `setindex!'
; ││││┌ @ pointer.jl:118 within `unsafe_store!'
        vmovsd  %xmm1, (%r14,%rbx,8)
; │└└└└
; │┌ @ range.jl:627 within `iterate'
; ││┌ @ promotion.jl:408 within `=='
        incq    %rbx
        cmpq    $4, %rbx
        jne     L80
```

it looks like it is reloading the whole tuple inside `c` in every inner iteration (64 times). Is it thinking that the store to the element in `c` could alias with the pointer to `c` itself..? It is something with the implementation of `Base.setindex!` (which is the same as in e.g. StaticArrays.jl).

**Comment by c42f** [comment]
I noticed this can be worked around by adding the following definition:

```julia
function Base.getindex(v::MMatrix4x4F64, i::Int)
    T = eltype(v)
    @assert isbitstype(T)
    @boundscheck checkbounds(v, i)
    GC.@preserve v unsafe_load(Base.unsafe_convert(Ptr{T}, pointer_from_objref(v)), i)
end
```

So yes, it's something to do with `getindex(c, i, k)` appearing in the right hand side of `c[i, k] += a[i, j] * b_jk` after lowering.

**Comment by KristofferC** [comment]
> I noticed this can be worked around by adding the following definition:

Cool, that takes it down to ~10 ns!
